client-js/app/elements/labrad-registry.ts

```diff
@@ -129,6 +129,101 @@ export class LabradRegistry extends polymer.Base {
     );
   }
 
+  //Drag and drop
+  @listen('dragenter')
+  overBehaviour(event) {
+    event.preventDefault(); //I don't understand this.
+  }
+
+  @listen('dragover')
+  changeBehaviour(event) {
+    //allows the ctrl key to be pressed to change cursor between copy/move
+    event.preventDefault(); //I don't understand this. But needs to be here
+
+    if (event.ctrlKey) {
+      event.dataTransfer.dropEffect = "copy";
+    }
+    else {
+      event.dataTransfer.dropEffect = "move";
+    }
+  }
+
+  @listen('dragleave')
+  leaveDrag(event) {
+    event.preventDefault(); //I don't understand this. But it needs to be here
+  }
+
+  @listen('dragstart')
+  startDrag(event) {
+    //detect start of drag event, grab info about target
+    var data: any; //I tried enum, but kept getting errors...
+    data = {path: this.path, name: event.target.name};
+    event.dataTransfer.setData('text', JSON.stringify(data));
+    event.dataTransfer.effectAllowed = 'copyMove';
+    this.$.toastText.text = "drag to move, hold ctrl to copy";
+    this.$.toastText.show();
+  }
+
+  @listen('dragend')
+  endDrag(event) {
+    console.log('drag ended',event);
+  }
+
+  dirDrop(event) {
+    //behaviour for dropping on folders
+    console.log(event);
+    var data: any;
+    data = JSON.parse(event.dataTransfer.getData('text'));
+    console.log('dirDrop');
+    event.target.closest('td').classList.remove('over');
+    if (event.ctrlKey) {
+      console.log('dropped on', event.target.closest('td').name, 'with data', data ); 
+      console.log(event.target);
+      var newPath: string[] = this.path.splice(0);
+      var oldFullPath: string[] = data.path;
```

Comment:
I have been using `Array<string>` for array types, rather than `string[]`. That's probably a residual scalaism, but in any case we should be consistent. The reason I prefer the former is that it looks like any other generic type that we might create for custom classes, such as `Observable<T>`, even if it's a bit more verbose. But I could be persuaded.

Of course in this case explicit types are not needed because the compiler will infer them, but it can still sometimes be useful for documentation.